<atomic>: atomic_ref<int8_t>::operator--(int) fails to compile

There is an incorrect cast on line 1368 of &lt;atomic&gt;. The cast to _Ty should be _TVal. 

    _TVal operator--(int) noexcept {
        return static_cast<_Ty>(_InterlockedExchangeAdd8(_Atomic_address_as<char>(this->_Storage), -1));
    }

**Command-line test case**
```
C:\Temp>type repro.cpp
#include <atomic>

int main() {
	int8_t x = 0;
	std::atomic_ref(x)--;
}

C:\Temp>cl /c /std:c++latest .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29304.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

repro.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29304\include\atomic(1368): error C2440: 'static_cast': cannot convert from 'char' to '_Ty'
        with
        [
            _Ty=int8_t &
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29304\include\atomic(1368): note: static_cast and safe_cast to reference can only be used for valid initializations or for lvalue casts between related classes
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29304\include\atomic(1367): note: while compiling class template member function 'signed char std::_Atomic_integral<_Ty &,1>::operator --(int) noexcept'
        with
        [
            _Ty=int8_t
        ]
repro.cpp(5): note: see reference to function template instantiation 'signed char std::_Atomic_integral<_Ty &,1>::operator --(int) noexcept' being compiled
        with
        [
            _Ty=int8_t
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29304\include\atomic(1781): note: see reference to class template instantiation 'std::_Atomic_integral<_Ty &,1>' being compiled
        with
        [
            _Ty=int8_t
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29304\include\atomic(2328): note: see reference to class template instantiation 'std::_Atomic_integral_facade<_Ty &>' being compiled
        with
        [
            _Ty=int8_t
        ]
repro.cpp(5): note: see reference to class template instantiation 'std::atomic_ref<int8_t>' being compiled
```

**Expected behavior**
Should compile successfully.

**STL version**
Latest

--- a/stl/inc/atomic
+++ b/stl/inc/atomic
@@ -1810,6 +1810,22 @@ struct _Atomic_integral_facade<_Ty&> : _Atomic_integral<_Ty&> {
         return fetch_add(_Negate(_Operand), _Order);
     }
 
+    _Ty operator++(int) const noexcept {
+        return const_cast<_Atomic_integral_facade*>(this)->_Base::operator++(0);
+    }
+
+    _Ty operator++() const noexcept {
+        return const_cast<_Atomic_integral_facade*>(this)->_Base::operator++();
+    }
+
+    _Ty operator--(int) const noexcept {
+        return const_cast<_Atomic_integral_facade*>(this)->_Base::operator--(0);
+    }
+
+    _Ty operator--() const noexcept {
+        return const_cast<_Atomic_integral_facade*>(this)->_Base::operator--();
+    }
+
     _Ty operator+=(const _Ty _Operand) const noexcept {
         return static_cast<_Ty>(fetch_add(_Operand) + _Operand);
     }

--- a/stl/inc/atomic
+++ b/stl/inc/atomic
@@ -1365,7 +1365,7 @@ struct _Atomic_integral<_Ty, 1> : _Atomic_storage<_Ty> { // atomic integral oper
     }
 
     _TVal operator--(int) noexcept {
-        return static_cast<_Ty>(_InterlockedExchangeAdd8(_Atomic_address_as<char>(this->_Storage), -1));
+        return static_cast<_TVal>(_InterlockedExchangeAdd8(_Atomic_address_as<char>(this->_Storage), -1));
     }
 
     _TVal operator--() noexcept {

--- a/tests/std/tests/P0019R8_atomic_ref/test.cpp
+++ b/tests/std/tests/P0019R8_atomic_ref/test.cpp
@@ -133,6 +133,14 @@ void test_int_ops() {
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x33);
     assert(ry.load() == 0x33);
+
+    vx--;
+    rx--;
+
+    assert(vx.load() == 0x32);
+    assert(vy.load() == 0x40);
+    assert(rx.load() == 0x32);
+    assert(ry.load() == 0x32);
 }
 
 

--- a/tests/std/tests/P0019R8_atomic_ref/test.cpp
+++ b/tests/std/tests/P0019R8_atomic_ref/test.cpp
@@ -177,16 +177,16 @@ void test_float_ops() {
     const std::atomic_ref rx(v);
     const std::atomic_ref ry(v);
 
-    vx.fetch_add(0x10);
-    rx.fetch_add(0x10);
+    assert(vx.fetch_add(0x10) == 0x40);
+    assert(rx.fetch_add(0x10) == 0x40);
 
     assert(vx.load() == 0x50);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x50);
     assert(ry.load() == 0x50);
 
-    vx.fetch_sub(0x8);
-    rx.fetch_sub(0x8);
+    assert(vx.fetch_sub(0x8) == 0x50);
+    assert(rx.fetch_sub(0x8) == 0x50);
 
     assert(vx.load() == 0x48);
     assert(vy.load() == 0x40);
@@ -212,16 +212,16 @@ void test_ptr_ops() {
     const std::atomic_ref rx(v);
     const std::atomic_ref ry(v);
 
-    vx.fetch_add(0x10);
-    rx.fetch_add(0x10);
+    assert(vx.fetch_add(0x10) == a);
+    assert(rx.fetch_add(0x10) == a);
 
     assert(vx.load() == a + 0x10);
     assert(vy.load() == a);
     assert(rx.load() == a + 0x10);
     assert(ry.load() == a + 0x10);
 
-    vx.fetch_sub(0x8);
-    rx.fetch_sub(0x8);
+    assert(vx.fetch_sub(0x8) == a + 0x10);
+    assert(rx.fetch_sub(0x8) == a + 0x10);
 
     assert(vx.load() == a + 0x8);
     assert(vy.load() == a);
@@ -230,6 +230,38 @@ void test_ptr_ops() {
 
     vx.store(a + 0x10);
     rx.store(a + 0x10);
+
+    assert(vx.load() == a + 0x10);
+    assert(vy.load() == a);
+    assert(rx.load() == a + 0x10);
+    assert(ry.load() == a + 0x10);
+
+    assert(vx-- == a + 0x10);
+    assert(rx-- == a + 0x10);
+
+    assert(vx.load() == a + 0xF);
+    assert(vy.load() == a);
+    assert(rx.load() == a + 0xF);
+    assert(ry.load() == a + 0xF);
+
+    assert(--vx == a + 0xE);
+    assert(--rx == a + 0xE);
+
+    assert(vx.load() == a + 0xE);
+    assert(vy.load() == a);
+    assert(rx.load() == a + 0xE);
+    assert(ry.load() == a + 0xE);
+
+    assert(vx++ == a + 0xE);
+    assert(rx++ == a + 0xE);
+
+    assert(vx.load() == a + 0xF);
+    assert(vy.load() == a);
+    assert(rx.load() == a + 0xF);
+    assert(ry.load() == a + 0xF);
+
+    assert(++vx == a + 0x10);
+    assert(++rx == a + 0x10);
 
     assert(vx.load() == a + 0x10);
     assert(vy.load() == a);

--- a/tests/std/tests/P0019R8_atomic_ref/test.cpp
+++ b/tests/std/tests/P0019R8_atomic_ref/test.cpp
@@ -134,13 +134,37 @@ void test_int_ops() {
     assert(rx.load() == 0x33);
     assert(ry.load() == 0x33);
 
-    vx--;
-    rx--;
+    assert(vx-- == 0x33);
+    assert(rx-- == 0x33);
 
     assert(vx.load() == 0x32);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x32);
     assert(ry.load() == 0x32);
+
+    assert(--vx == 0x31);
+    assert(--rx == 0x31);
+
+    assert(vx.load() == 0x31);
+    assert(vy.load() == 0x40);
+    assert(rx.load() == 0x31);
+    assert(ry.load() == 0x31);
+
+    assert(vx++ == 0x31);
+    assert(rx++ == 0x31);
+
+    assert(vx.load() == 0x32);
+    assert(vy.load() == 0x40);
+    assert(rx.load() == 0x32);
+    assert(ry.load() == 0x32);
+
+    assert(++vx == 0x33);
+    assert(++rx == 0x33);
+
+    assert(vx.load() == 0x33);
+    assert(vy.load() == 0x40);
+    assert(rx.load() == 0x33);
+    assert(ry.load() == 0x33);
 }
 
 

--- a/tests/std/tests/P0019R8_atomic_ref/test.cpp
+++ b/tests/std/tests/P0019R8_atomic_ref/test.cpp
@@ -94,40 +94,40 @@ void test_int_ops() {
     const std::atomic_ref rx(v);
     const std::atomic_ref ry(v);
 
-    vx.fetch_add(0x10);
-    rx.fetch_add(0x10);
+    assert(vx.fetch_add(0x10) == 0x40);
+    assert(rx.fetch_add(0x10) == 0x40);
 
     assert(vx.load() == 0x50);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x50);
     assert(ry.load() == 0x50);
 
-    vx.fetch_sub(0x8);
-    rx.fetch_sub(0x8);
+    assert(vx.fetch_sub(0x8) == 0x50);
+    assert(rx.fetch_sub(0x8) == 0x50);
 
     assert(vx.load() == 0x48);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x48);
     assert(ry.load() == 0x48);
 
-    vx.fetch_or(0xF);
-    rx.fetch_or(0xF);
+    assert(vx.fetch_or(0xF) == 0x48);
+    assert(rx.fetch_or(0xF) == 0x48);
 
     assert(vx.load() == 0x4F);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0x4F);
     assert(ry.load() == 0x4F);
 
-    vx.fetch_and(0x3C);
-    rx.fetch_and(0x3C);
+    assert(vx.fetch_and(0x3C) == 0x4F);
+    assert(rx.fetch_and(0x3C) == 0x4F);
 
     assert(vx.load() == 0xC);
     assert(vy.load() == 0x40);
     assert(rx.load() == 0xC);
     assert(ry.load() == 0xC);
 
-    vx.fetch_xor(0x3F);
-    rx.fetch_xor(0x3F);
+    assert(vx.fetch_xor(0x3F) == 0xC);
+    assert(rx.fetch_xor(0x3F) == 0xC);
 
     assert(vx.load() == 0x33);
     assert(vy.load() == 0x40);

